tempesta_fw/sock.c

```diff
@@ -990,7 +994,12 @@ ss_tcp_state_change(struct sock *sk)
 		if (!skb_queue_empty(&sk->sk_receive_queue))
 			ss_tcp_process_data(sk);
 		TFW_DBG2("[%d]: Peer connection closing\n", smp_processor_id());
-		ss_linkerror(sk);
+		/*
+		 * Use ss_close() to close the connection after
+		 * sending responses in the queue, for example,
+		 * tls alerts
+		 */
+		ss_close(sk, SS_F_SYNC);
```

Comment:
Let's see, what is really happen here. There could be a few cases:

- Backend server has sent us a TCP frame with `FIN` flag set. It means that it has replied everything it wanted, there is no any good to send it more requests: it'll never reply to them. So it's better for us to immediately close the connection and avoid sending more requests to it.

- Client has sent us `FIN` in last request inside http connection. I don't remember such situation in the RFC, but it seems  that here is no need to process requests in the receive queue at all: any way it is to be closed before response from backend server will be received. In this case we should either close client immediately without receive_queue processing, or send all the responses and close the connection after last is sent.

- Client has sent us `FIN` inside HTTPS connection and we have alerts to send - yes, it seams reasonable to perform a scheduled connection close by `ss_close()` call.

I would say that replacing `ss_linkerror()` to `ss_close()` is mostly performance question here.

Reply:
@krizhanovsky

Reply:
Is there a clear comment why `ss_linkerror()` is no longer adequate in `ss_tx_action()` and `ss_close()` is preferred there? I am not sure I understand the reason for the change. Thanks.

Reply:
This is `ss_tcp_state_change()`, not `ss_tx_action()`.
We do not want to close the connection immediately, but only after sending the responses in the queue. For example, TLS alerts. See comment above.

Reply:
I am sorry that I am not familiar with the TLS part.

I am not sure which connection and which side of the connection is in question here. This particular place is where TCP state change occurs when the other side has sent `FIN`  - initiated the closing of the connection. After that, the other side does not care of receiving any data. Perhaps, that's not the case for a TLS connection, and the other side, the one that has sent `FIN`, expects to receive something after that? That would be strange in terms of a TCP connection, in my opinion.  Perhaps, a better comment is necessary.

Now, the receive queue is processed here because the packet with `FIN` might not be the only one in the queue but be the last one. Data in those preceding packets, including the packet with `FIN`, need to be processed by Tempesta. This place is called before `ss_tcp_data_ready()`, and so we just process the receive queue and close the connection directly, thus making it so `ss_tcp_data_ready()` is not called at all.

If  we're talking about doing something (like sending data to the other side of a _live_ connection) _before_ sending FIN ourselves (before _initiating_ the closing of a connection), then perhaps the place in `ss_tcp_state_change()` is not the right place for that. This place is about handling a connection closing initiated by the other side.

It's very similar in the other function `ss_tcp_data_ready()`, the logic is very much the same. There it's more about internal Tempesta errors though. Note that nearly all errors are handled within Tempesta itself by performing some actions and initiating a connection closing (usually, a scheduled closing). It's a rare case an error is returned from Tempesta to the socket layer to `ss_tcp_data_ready()`. Those are very specific cases of "hard" errors where that is justified, and so `ss_linkerror()` is the right choice there. If an error is returned from TLS layer to the socket layer in a case that needs additional handling while the socket is alive, then, perhaps, that is something that should not happen, by the overall design of Tempesta.

(_The comment about FIN in `ss_tcp_data_ready()` there may no longer be true now, need to check that. The socket should be dead after closing in `ss_tcp_state_change()`, and so `ss_tcp_data_ready() is not called. Also, all skb are flushed from the queue in the process anyway._)

The difference  between `ss_linkerror()` and `ss_close()` is two-fold.

- Different Tempesta callback is called, `connection_error()` versus `connection_drop()`
- One is direct closing, and the other is a scheduled, postponed closing.

In both cases a callback into Tempesta is called _after_ the connection is closed. In that regard, the connection _is_ closed "immediately", and not after something else is done in a callback. The only very subtle difference is the "postpone" part, where the closing action is performed in a specific SoftIRQ (different from RX/TX SoftIRQ), and therefore there _may_ be a "cycle" of Tempesta work before that specific SoftIRQ kicks in. But that's not guaranteed - this  scheduling of the closing action may be the last work for RX/TX SoftIRQ, and then the specific  SoftIRQ would kick in immediately.

All in all, I believe this issue would greatly benefit from a clear description of the scenario (or several scenarios) in mind, and how the changes help in making those scenarios work correctly versus the existing code. Perhaps, describe that in the issue on Github. Right now it's difficult to see how these changes fix the issue.

Reply:
> Backend server has sent us a TCP frame with FIN flag set. It means that it has replied everything it wanted, there is no any good to send it more requests: it'll never reply to them. So it's better for us to immediately close the connection and avoid sending more requests to it.

Good point: we should take advantage on knowledge that at TCP layer we know about application layer and reduce amount of work to do. It could make sense to leave current `ss_linkerror()` for server connections, introducing additional knowledge of the socket type on SS layer (now we don't know in sock.c which sockets are server and which are client).

However, with that we've been living happily with the code until TLS in mind, this proposal seems error prone. We'll have full TLS proxying (so, we'll have to send alerts to backends as well) and also there will be websockets with full duplex connection and, the worse, we don't plan to understand websockets protocol. So at this point we can make the optimization for HTTP server connections only. Maybe we'll see in future a case with massive FINs from HTTP backends, but I don't want to do the optimization now.

But I believe the opportunity must be commented.

> Client has sent us FIN in last request inside http connection. I don't remember such situation in the RFC, but it seems that here is no need to process requests in the receive queue at all: any way it is to be closed before response from backend server will be received.

This is allowed and adequate for a client to send a request with `Connection: close` and add FIN to the same TCP segment. It seems we simply have no test cases for the scenario, but I believe #1308 also affect the scenario.

`ss_linkerror()`, called on receiving FIN from a peer, calls `ss_do_close()` with immediate FIN transmission to the peer regardless some application data in our work queue which is still was not added to TCP send queue. This breaks TCP FSM: on receiving FIN we have to reply ACK and more to CLOSE WAIT state where we send all required data to the peer and only after than send our FIN. An application layer, as we see with TLS now and probably see with client request with FIN and `Connection: close`, may be broken by our current behaviour.